Readme.md
# retext-cst [![Build Status](https://img.shields.io/travis/wooorm/retext-cst.svg?style=flat)](https://travis-ci.org/wooorm/retext-cst) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-cst.svg?style=flat)](https://coveralls.io/r/wooorm/retext-cst?branch=master)

**[retext](https://github.com/wooorm/retext "Retext")** encoding and decoding between [NLCST](https://github.com/wooorm/nlcst) and object model.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-cst
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-cst
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-cst
```

[Duo](http://duojs.org/#getting-started):

```javascript
var cst = require('wooorm/retext-cst');
```

## Usage

```javascript
var Retext = require('retext');
var cst = require('retext-cst');
var inspect = require('retext-inspect');
var fs = require('fs');

/* Read a JSON stringified NLCST document. */
var nlcst = fs.readFileSync('cst.json', 'utf8');

var retext = new Retext()
    .use(inspect)
    .use(cst);

retext.fromCST(nlcst, function (err, tree) {
    if (err) throw err;

    /*
     * Pretty-print the object model as a cst (each
     * level with two spaces).
     */

    console.log(tree.toCST(2));
});
```

## API

### [retext](https://github.com/wooorm/retext#retextparser).fromCST(cst, options?, done(err, tree))

Parse an [NLCST node](https://github.com/wooorm/nlcst#node) (stringified or JSON) into an object model.

- `cst` (`Object` or `string`): The object to parse into a TextOM tree;
- `options` (`Object`): Passed to plugins;
- `done` (`function(Error, TextOMNode)`): callback.

Let’s say we have the following (abbreviated) NCLST in a file called `cst.json`.

```json
{
  "type": "RootNode",
  "children": [
    {
      "type": "ParagraphNode",
      "children": [
        {
          "type": "SentenceNode",
          "children": [
            {
              "type": "WordNode",
              "children": [
                {
                  "type": "TextNode",
                  "value": "A"
                }
              ]
            },
            ...
            {
              "type": "PunctuationNode",
              "value": "."
            }
          ]
        }
      ]
    }
  ]
}
```

Such a file would result in the following **retext** document (note that [retext-inspect](https://github.com/wooorm/retext-inspect) is also used):

```javascript
retext.fromCST(theAboveJsonDocument, function (err, tree) {
    if (err) throw err;

    console.log(tree);
    /*
     * RootNode[1]
     * └─ ParagraphNode[1]
     *    └─ SentenceNode[2]
     *       ├─ WordNode[1]
     *       │  └─ TextNode: 'A'
     *       ...
     *       └─ PunctuationNode: '.'
     */
});
```

### [Node](https://github.com/wooorm/textom#textomnode-nlcstnode)#toJSON()

```javascript
retext.parse('A simple sentence.', function (err, tree) {
    tree.toJSON();
    /*
     * ˅ Object
     *    ˃ children: Array[1]
     *      type: "RootNode"
     *    ˃ __proto__: Object
     */
});
```

Returns a JSON (**not** stringified) representation of a Node, which can later be passed to [`retext.fromCST()`](#retextfromcstcst).

This is just a wrapper around `Node#valueOf()`[[1](https://github.com/wooorm/textom#textomparentvalueof), [2](https://github.com/wooorm/textom#textomtextvalueof)], so it’s possible to pass nodes to [`JSON.stringify`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON_behavior).

### [Node](https://github.com/wooorm/textom#textomnode-nlcstnode)#toCST(delimiter?)

```javascript
retext.parse('A simple sentence.', function (err, tree) {
    tree.toCST();
    /*
     * '{"type":"RootNode","children":[{"type":"ParagraphNode","children":[{"type":"SentenceNode","children":[{"type":"WordNode","children":[{"type":"TextNode","value":"A"}]},{"type":"WhiteSpaceNode","value":" "},{"type":"WordNode","children":[{"type":"TextNode","value":"simple"}]},{"type":"WhiteSpaceNode","value":" "},{"type":"WordNode","children":[{"type":"TextNode","value":"sentence"}]},{"type":"PunctuationNode","value":"."}]}]}]}'
     */
});
```

Returns a stringified JSON NLCST representation of a Node.

- `delimiter` (`null`, `number`, or `string`): Pretty prints the CST. Passed to [`JSON.stringify`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#space_argument).

## Performance

```text
              retext.fromCST(cst, options?, callback);
     258 op/s » A paragraph (5 sentences, 100 words)
      27 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
       3 op/s » An article (100 paragraphs, 500 sentences, 10,000 words)

              TextOM.Node#toJSON()
  26,037 op/s » A paragraph (5 sentences, 100 words)
   2,669 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
     223 op/s » An article (100 paragraphs, 500 sentences, 10,000 words)

              TextOM.Node#toCST(delimiter?)
   1,003 op/s » A paragraph (5 sentences, 100 words)
     101 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
       9 op/s » An article (100 paragraphs, 500 sentences, 10,000 words)
```

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.3.3 / 2015-01-13
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor API reference in `Readme.md`
  * Refactor code examples in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update eslint
  * Update nlcst-to-textom
  * Fix malformed `bower.json`
  * Update retext
  * Fix travis deploy authentication

0.3.2 / 2014-12-05
==================

 * Add benchmark
 * Fix non-github-markdown in `Readme.md`
 * Refactor npm script targets in `package.json`
 * Add link to personal website to copyright in `Readme.md`
 * Update nlcst-to-textom, eslint, retext

0.3.1 / 2014-11-20
==================

 * Fix incorrect names in package manifests

0.3.0 / 2014-11-20
==================

 * Rename project in `.travis.yml`
 * Add support for an optional `options` object to `fromCST`
 * Remove exposed `toJSON` and `toCST` methods
 * Refactor `index.js`
 * Add flat badges to `Readme.md`
 * Refactor `Readme.md`
 * Rename project
 * Fix incorrect executive rights on `test.js`
 * Refactor npm script targets in `package.json`
 * Update retext, nlcst-to-textom

0.2.2 / 2014-10-27
==================

 * Add `.eslintrc`
 * Refactor `.jscs.json`
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update eslint, mocha, retext

0.2.1 / 2014-10-21
==================

 * Refactor test
 * Update retext

0.2.0 / 2014-10-14
==================

 * Remove custom methods
 * Add nlcst-to-textom as a dependency
 * Update .gitignore, .npmignore, bower ignore
 * Move spec to test.js
 * Remove browser test
 * Fix error messages

0.1.2 / 2014-09-24
==================

 * Update Readme.md
 * Add passing errors to `done` instead of throwing
 * Refactor API
 * Add npm deployment to travis
 * Remove `npm update` from travis
 * Update Installation in docs
 * Add bower.json
 * Fix property order in component.json, package.json
 * Update .gitignore, .npmignore
 * Update copright in Readme.md
 * Remove testling
 * Refactor spec for retext@0.2.0
 * Update retext
 * Fix API for retext@0.2.0
 * Add retext as a peer dependency
 * Update eslint to 0.8.0, mocha to 1.20.4

0.1.1 / 2014-07-14
==================

 * fromAST now include data properties, when existing
 * Fix incorrect wording (ast instead of tree)
 * Fix incorrect wording
 * toJSON and toAST now include data properties, when not empty (fixes #1)

0.1.0 / 2014-07-12
==================

 * Update documentation for latest changes in retext
 * Added changelog
 * Simplified unit tests
 * Fix changed functionality in textom, retext
 * Update retext
 * Removed functionality to test complexity, browserify unit tests by default

0.0.4 / 2014-06-19
==================

c319e58 Unit tests can now run in the browser (spec/index.html)
f0b3f01 Fixed a bug where everything in ./spec/ was tested
d6f8c74 Split the lint task into lint-api, lint-test, and lint-style
7b52c3b Updated dependency versions of jscs and retext

0.0.3 / 2014-06-17
==================

c32386a Fix code styleguide for latest jscs updates
ef09b25 Update dependency versions of jscs and retext
4cddff5 Update retext developer dependency to 0.0.10
9b2da12 Update retext developer dependency to 0.0.9

0.0.2 / 2014-06-13
==================

90364f0 Fixed Readme

0.0.1 / 2014-06-13
==================


test.js
'use strict';

/*
 * Dependencies.
 */

var retextCST,
    Retext,
    assert;

retextCST = require('./');
Retext = require('retext');
assert = require('assert');

/*
 * Retext.
 */

var retext,
    TextOM,
    parser;

retext = new Retext().use(retextCST);
TextOM = retext.TextOM;
parser = retext.parser;

/*
 * Tests.
 */

describe('retextCST()', function () {
    it('should be a `function`', function () {
        assert(typeof retextCST === 'function');
    });
});

describe('retextCST.attach()', function () {
    it('should attach a `toCST` method to `TextOM.Node#`', function () {
        assert(typeof TextOM.Node.prototype.toCST === 'function');
    });

    it('should attach a `toJSON` method to `TextOM.Node#`', function () {
        assert(typeof TextOM.Node.prototype.toJSON === 'function');
    });

    it('should attach a `fromCST` method to `retext`', function () {
        assert(typeof retext.fromCST === 'function');
    });
});

describe('TextOM.Node#toJSON()', function () {
    it('should be a `function`', function () {
        assert(typeof (new TextOM.RootNode()).toJSON === 'function');
    });

    it('should convert a `Node` into an CST', function (done) {
        var source,
            cst;

        source = 'A simple sentence.';
        cst = parser.parse(source);

        retext.parse(source, function (err, tree) {
            assert(JSON.stringify(tree) === JSON.stringify(cst));

            done(err);
        });
    });

    it('should throw, when not operating on a Node', function () {
        assert.throws(function () {
            new TextOM.Node().toJSON.call(Math);
        }, /Math/);

        assert.throws(function () {
            new TextOM.Node().toJSON.call();
        }, /undefined/);
    });

    it('should include a data property when non-empty', function (done) {
        var source,
            cst;

        source = 'A simple sentence.';
        cst = parser.parse(source);

        cst.data = {
            'test': 'test'
        };

        retext.parse(source, function (err, tree) {
            tree.data.test = 'test';

            assert(JSON.stringify(tree) === JSON.stringify(cst));

            done(err);
        });
    });
});

describe('TextOM.Node#toCST(delimiter?)', function () {
    it('should be a `function`', function () {
        assert(typeof new TextOM.RootNode().toCST === 'function');
    });

    it('should convert a `Node` into an stringified CST', function (done) {
        var source,
            cst;

        source = 'A simple sentence.';
        cst = parser.parse(source);

        retext.parse(source, function (err, tree) {
            assert(tree.toCST() === JSON.stringify(cst));

            done(err);
        });
    });

    it('should convert a `Node` into an stringified CST using the given ' +
        'delimiter', function (done) {
            var source,
                delimiter = '\t',
                cst;

            source = 'A simple sentence.';
            cst = parser.parse(source);

            retext.parse(source, function (err, tree) {
                assert(
                    tree.toCST(delimiter) ===
                    JSON.stringify(cst, null, delimiter)
                );

                done(err);
            });
        }
    );
});

describe('Retext.fromCST(cst, options?, done)', function () {
    it('should be a `function`', function () {
        assert(typeof retext.fromCST === 'function');
    });

    it('should throw when `done` is omitted', function () {
        assert.throws(function () {
            retext.fromCST({
                'type': 'TextNode',
                'value': 'A'
            });
        }, /not a valid argument for/);
    });

    it('should throw, when something other than a string or object is given',
        function (done) {
            retext.fromCST(Math, function (err) {
                assert.throws(function () {
                    throw err;
                }, /object Math/);
            });

            done();
        }
    );

    it('should NOT throw when `options` are given', function (done) {
        retext.fromCST({
            'type': 'TextNode',
            'value': 'A'
        }, {}, done);
    });

    it('should throw, when the `JSON.Parse`d value does not contain a ' +
        '`type` attribute', function (done) {
            retext.fromCST({
                'value': 'test'
            }, function (err) {
                assert.throws(function () {
                    throw err;
                });

                done();
            });
        }
    );

    it('should throw, when the `JSON.Parse`d value contains ' +
        'neither a `children`, nor a `value` attribute',
        function (done) {
            retext.fromCST({
                'type': 'RootNode'
            }, function (err) {
                assert.throws(function () {
                    throw err;
                });

                done();
            });
        }
    );

    it('should convert a stringified CST into an object model',
        function (done) {
            var cst;

            cst = JSON.stringify(parser.parse('A simple sentence.'));

            retext.fromCST(cst, function (err, tree) {
                assert(tree.toCST() === cst);

                done(err);
            });
        }
    );

    it('should accept `String` objects',
        function (done) {
            var cst;

            cst = JSON.stringify(parser.parse('A simple sentence.'));

            /* eslint-disable no-new-wrappers */
            retext.fromCST(new String(cst), function (err, tree) {
                assert(tree.toCST() === cst);

                done(err);
                /* eslint-enable no-new-wrappers */
            });
        }
    );

    it('should convert an CST into an object model', function (done) {
        var cst;

        cst = parser.parse('A simple sentence.');

        retext.fromCST(cst, function (err, tree) {
            assert(tree.toCST() === JSON.stringify(cst));

            done(err);
        });
    });

    it('should set data properties', function (done) {
        var cst;

        cst = parser.parse('A simple sentence.');

        cst.data = {
            'test': 'test'
        };

        cst = JSON.stringify(cst);

        retext.fromCST(cst, function (err, tree) {
            assert(tree.toCST() === cst);

            done(err);
        });
    });
});


spec/retext-ast.spec.js
'use strict';

var retextAST,
    Retext,
    assert,
    retext,
    TextOM,
    parser;

/**
 * Module dependencies.
 */

retextAST = require('..');
Retext = require('retext');
assert = require('assert');

/**
 * Retext.
 */

retext = new Retext().use(retextAST);
TextOM = retext.TextOM;
parser = retext.parser;

/**
 * Tests.
 */

describe('retextAST()', function () {
    it('should be a `function`', function () {
        assert(typeof retextAST === 'function');
    });

    it('should export an `attach` method', function () {
        assert(typeof retextAST.attach === 'function');
    });

    it('should export a `toJSON` method', function () {
        assert(typeof retextAST.toJSON === 'function');
    });

    it('should export a `toAST` method', function () {
        assert(typeof retextAST.toAST === 'function');
    });
});

describe('retextAST.attach()', function () {
    it('should attach a `toAST` method to `TextOM.Node#`', function () {
        assert(typeof TextOM.Node.prototype.toAST === 'function');
    });

    it('should attach a `toJSON` method to `TextOM.Node#`', function () {
        assert(typeof TextOM.Node.prototype.toJSON === 'function');
    });

    it('should attach a `fromAST` method to `retext`', function () {
        assert(typeof retext.fromAST === 'function');
    });
});

describe('TextOM.Node#toJSON()', function () {
    it('should be a `function`', function () {
        assert(typeof (new TextOM.RootNode()).toJSON === 'function');
    });

    it('should convert a `Node` into an AST', function (done) {
        var source,
            ast;

        source = 'A simple sentence.';
        ast = parser.parse(source);

        retext.parse(source, function (err, tree) {
            assert(JSON.stringify(tree) === JSON.stringify(ast));

            done(err);
        });
    });

    it('should throw, when not operating on a Node', function () {
        assert.throws(function () {
            new TextOM.Node().toJSON.call(Math);
        }, /Math/);

        assert.throws(function () {
            new TextOM.Node().toJSON.call();
        }, /undefined/);
    });

    it('should include a data property when non-empty', function (done) {
        var source,
            ast;

        source = 'A simple sentence.';
        ast = parser.parse(source);

        ast.data = {
            'test' : 'test'
        };

        retext.parse(source, function (err, tree) {
            tree.data.test = 'test';

            assert(JSON.stringify(tree) === JSON.stringify(ast));

            done(err);
        });
    });
});

describe('TextOM.Node#toAST(delimiter?)', function () {
    it('should be a `function`', function () {
        assert(typeof new TextOM.RootNode().toAST === 'function');
    });

    it('should convert a `Node` into an stringified AST', function (done) {
        var source,
            ast;

        source = 'A simple sentence.';
        ast = parser.parse(source);

        retext.parse(source, function (err, tree) {
            assert(tree.toAST() === JSON.stringify(ast));

            done(err);
        });
    });

    it('should convert a `Node` into an stringified AST using the given ' +
        'delimiter', function (done) {
            var source,
                delimiter = '\t',
                ast;

            source = 'A simple sentence.';
            ast = parser.parse(source);

            retext.parse(source, function (err, tree) {
                assert(
                    tree.toAST(delimiter) ===
                    JSON.stringify(ast, null, delimiter)
                );

                done(err);
            });
        }
    );
});

describe('Retext.fromAST(ast, done)', function () {
    it('should be a `function`', function () {
        assert(typeof retext.fromAST === 'function');
    });

    it('should throw, when something other than a string or object is given',
        function (done) {
            retext.fromAST(Math, function (err) {
                assert.throws(function () {
                    throw err;
                }, /object Math/);
            });

            done();
        }
    );

    it('should throw, when the `JSON.Parse`d value does not contain a ' +
        '`type` attribute', function (done) {
            retext.fromAST({
                'value' : 'test'
            }, function (err) {
                assert.throws(function () {
                    throw err;
                }, /`type`/);

                done();
            });
        }
    );

    it('should throw, when the `JSON.Parse`d value contains ' +
        'neither a `children`, nor a `value` attribute',
        function (done) {
            retext.fromAST({
                'type' : 'RootNode'
            }, function (err) {
                assert.throws(function () {
                    throw err;
                }, /`(value|children)`/);

                done();
            });
        }
    );

    it('should convert a stringified AST into an object model',
        function (done) {
            var ast;

            ast = JSON.stringify(parser.parse('A simple sentence.'));

            retext.fromAST(ast, function (err, tree) {
                assert(tree.toAST() === ast);

                done(err);
            });
        }
    );

    it('should accept `String` objects',
        function (done) {
            var ast;

            ast = JSON.stringify(parser.parse('A simple sentence.'));

            /* eslint-disable no-new-wrappers */
            retext.fromAST(new String(ast), function (err, tree) {
                assert(tree.toAST() === ast);

                done(err);
                /* eslint-enable no-new-wrappers */
            });
        }
    );

    it('should convert an AST into an object model', function (done) {
        var ast;

        ast = parser.parse('A simple sentence.');

        retext.fromAST(ast, function (err, tree) {
            assert(tree.toAST() === JSON.stringify(ast));

            done(err);
        });
    });

    it('should set data properties', function (done) {
        var ast;

        ast = parser.parse('A simple sentence.');

        ast.data = {
            'test' : 'test'
        };

        ast = JSON.stringify(ast);

        retext.fromAST(ast, function (err, tree) {
            assert(tree.toAST() === ast);

            done(err);
        });
    });
});


index.js
'use strict';

/*
 * Dependencies.
 */

var toTextOM;

toTextOM = require('nlcst-to-textom');

/*
 * Warn when this plug-in will not work: JSON encoding
 * and decoding is required.
 */

/* istanbul ignore if */
if (!JSON) {
    throw new Error(
        'Missing `JSON` for `retext-cst`'
    );
}

/**
 * Transform an NLCST node into a TextOM node.
 *
 * @param {Object} TextOM - Object model.
 * @param {NLCSTNode|string} nlcst - CST node to
 *   transform.
 * @return {TextOMNode}
 */
function fromJSON(TextOM, nlcst) {
    var node;

    if (nlcst instanceof String) {
        nlcst = nlcst.toString();
    }

    if (typeof nlcst === 'string') {
        nlcst = JSON.parse(nlcst);
    }

    try {
        node = toTextOM(TextOM, nlcst);
    } catch (exception) {
        throw new Error(
            'Illegal invocation: `' + nlcst + '` ' +
            'is not a valid argument for `fromJSON`'
        );
    }

    return node;
}

/**
 * Transform an NLCST node into a TextOM node, and
 * run plugins on it.
 *
 * @this {Retext}
 * @param {NLCSTNode|string} nlcst - The concrete syntax tree to
 *   transform.
 * @param {Object?} options
 * @param {function(Error, Node)} done - Callback to
 *   invoke when the transformations have completed.
 * @return {Retext} - Self.
 */
function fromCST(nlcst, options, done) {
    var self,
        tree;

    self = this;

    if (!done) {
        done = options;
        options = null;
    }

    if (typeof done !== 'function') {
        throw new TypeError(
            'Illegal invocation: `' + done + '` ' +
            'is not a valid argument for ' +
            '`Retext#fromCST(value, done)`.\n' +
            'This breaking change occurred in 0.2.0, ' +
            'see GitHub for more information.'
        );
    }

    try {
        tree = fromJSON(self.TextOM, nlcst);
    } catch (err) {
        done(err);

        return self;
    }

    self.run(tree, done);

    return self;
}

/**
 * Transform a TextOM node into an NLCST node.
 *
 * @this {TextOMNode}
 * @return {NLCSTNode}
 */
function toJSON() {
    var self;

    self = this;

    if (!self || !self.TextOM) {
        throw new Error(
            'Illegal invocation: `' + self + '` ' +
            'is not a valid context for `toJSON`'
        );
    }

    return self.valueOf();
}

/**
 * Transform a TextOM node into a stringified NLCST node.
 *
 * @this {TextOMNode}
 * @param {string|number?} delimiter - When given,
 *   pretty prints the stringified object—indenting
 *   each level either with the given string or with
 *   the given number of spaces.
 * @return {string}
 */
function toCST(delimiter) {
    return JSON.stringify(this, null, delimiter);
}

/**
 * Define `retextCST`.
 *
 * @param {Retext} retext - Instance of Retext.
 */
function retextCST(retext) {
    var nodePrototype;

    nodePrototype = retext.TextOM.Node.prototype;

    nodePrototype.toCST = toCST;
    nodePrototype.toJSON = toJSON;

    retext.fromCST = fromCST;
}

/*
 * Expose `retextCST`.
 */

module.exports = retextCST;


benchmark.js
'use strict';

var Retext,
    cst;

/*
 * Dependencies.
 */

Retext = require('retext');
cst = require('./');

/*
 * Dependencies.
 */

var retext;

retext = new Retext().use(cst);

/*
 * Test data.
 *
 * This includes:
 *
 * - An average sentence (w/ 20 words);
 * - An average paragraph (w/ 5 sentences);
 * - A (big?) section (w/ 10 paragraphs);
 * - A (big?) article (w/ 10 sections);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section,
    article,
    paragraphNLCST,
    sectionNLCST,
    articleNLCST;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

article = section + Array(10).join('\n\n' + section);

paragraphNLCST = retext.parser.parse(paragraph);
sectionNLCST = retext.parser.parse(section);
articleNLCST = retext.parser.parse(article);

/*
 * Benchmark `fromCST`.
 */

suite('retext.fromCST(cst, options?, callback);', function () {
    bench('A paragraph (5 sentences, 100 words)',
        function (done) {
            retext.fromCST(paragraphNLCST, done);
        }
    );

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retext.fromCST(sectionNLCST, done);
        }
    );

    bench('An article (100 paragraphs, 500 sentences, 10,000 words)',
        function (done) {
            retext.fromCST(articleNLCST, done);
        }
    );
});

/*
 * Benchmark `toJSON`.
 */

suite('TextOM.Node#toJSON()', function () {
    var paragraphNode,
        sectionNode,
        articleNode;

    before(function (done) {
        retext.fromCST(paragraphNLCST, function (err, tree) {
            paragraphNode = tree;

            done(err);
        });
    });

    before(function (done) {
        retext.fromCST(sectionNLCST, function (err, tree) {
            sectionNode = tree;

            done(err);
        });
    });

    before(function (done) {
        retext.fromCST(articleNLCST, function (err, tree) {
            articleNode = tree;

            done(err);
        });
    });

    bench('A paragraph (5 sentences, 100 words)', function () {
        paragraphNode.toJSON();
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function () {
            sectionNode.toJSON();
        }
    );

    bench('An article (100 paragraphs, 500 sentences, 10,000 words)',
        function () {
            articleNode.toJSON();
        }
    );
});

/*
 * Benchmark `toCST`.
 */

suite('TextOM.Node#toCST(delimiter?)', function () {
    var paragraphNode,
        sectionNode,
        articleNode;

    before(function (done) {
        retext.fromCST(paragraphNLCST, function (err, tree) {
            paragraphNode = tree;

            done(err);
        });
    });

    before(function (done) {
        retext.fromCST(sectionNLCST, function (err, tree) {
            sectionNode = tree;

            done(err);
        });
    });

    before(function (done) {
        retext.fromCST(articleNLCST, function (err, tree) {
            articleNode = tree;

            done(err);
        });
    });

    bench('A paragraph (5 sentences, 100 words)', function () {
        paragraphNode.toCST();
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function () {
            sectionNode.toCST();
        }
    );

    bench('An article (100 paragraphs, 500 sentences, 10,000 words)',
        function () {
            articleNode.toCST();
        }
    );
});
